### PR TITLE
#21 Introduce Email Service

### LINKED ISSUE #21: Set-up Email Service
[//]: # (These are comments and are used for reference and will not show up in the issue)

**Description:**
eVoter's main form of communication is through email and is used in various functionalities.

Preferable to use a no-reply email sender, but if we'll have to set up an email account for eVoter, that works too.

Don't need to write the actual email templates for each related issue. Just set up the base methods as a prerequisite to sending email functionality

[//]: # (A clear and concise description of what you want to happen.)

**Related Issues or Functional Requirement(s):**
#2, #5, #9 

[//]: # (Ex. #<issue id>. Moreover, please add the appropriate FR label.)

**Related WBS Block Number(s):**
N/A

**Additional context:**
Might be helpful - https://mailslurp.medium.com/sending-emails-in-javascript-3-ways-to-send-and-test-emails-with-nodejs-8f3e5c3d0964

[//]: # (Add any other context or screenshots about the feature request here.)

### PR DESCRIPTION
**Pull request by gatilogo**
[//]: # (These are comments and are used for reference and will not show up in the PR)

**Description:** 

[//]: # (Include a summary of the change and which issue is fixed. List any dependencies that are required for this change)

Setup email service in backend by adding Send Grid for sending email functionality. API key has been removed for security. I forgot to remove it in my first commit lol so I'll just generate a new one

Usage:
```
export SENDGRID_API_KEY=<Your API Key>
```
```
const msg = {
    intention:"Invitation",
    to: "someone@gmail.com", // Change to your recipient
    from: {
      email: "noreply.evoter@gmail.com",
      name: "eVoter Support",
    },
    subject: "Sending with SendGrid Test", // Change to your subject
    text: "Hello from eVoter Node.js app", // Body of email could be either text or html
    html: "<strong>Hello from eVoter Node.js app</strong>",
  };
const { validateEmail, sendEmail } =require("./services/mailer");
isSuccessful = sendEmail(msg);
```

Add-On:
- Expose mailer method which sends email
- Expose `validateEmail()` and `setKey()`
- Add email message validator
- Add extra field "Intention" to email
- Add email messages for sent and not sent
- fix some EOF warning
- move mailer to services folder

**Proof of concept:**
image

[//]: # (Replace with your issue id to automatically close given issue number upon closing PR)
Closes #21

### CODE REVIEW

**Review by LiyaoJiang1998 — APPROVED** [review]
email service looks good!

**Comment by gza060625** [comment]
Change mailer.js into one "service"
- Expose mailer method which sends email
- Expose `validateEmail()` and `setKey()`
- Add email message validator
- Add extra field "Intention" to email
- Add email messages for sent and not sent
- fix some EOF warning
- move mailer to services folder

Ask for review due to the changes @LiyaoJiang1998 
@gatilogo Couldn't add you as one reviewer, please have a look as well.

**Review by gza060625 — APPROVED** [review]
LGTM

File: backend/services/mailer.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+const sgMail = require("@sendgrid/mail");
+const Joi = require("joi");
+
+function setKey(){
+  sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+}
+setKey();
+const schema = Joi.object({
+    intention:Joi.string().valid('Invitation', 'F2A', 'Access Code').required(),
```
Comment (LiyaoJiang1998) [inline backend/services/mailer.js:9]:
```suggestion
    intention:Joi.string().valid('Invitation', '2FA', 'Access Code').required(),
```

**Review by gza060625 — DISMISSED** [review]

File: backend/services/mailer.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    to: Joi.string().email().required(),
+    from: Joi.object({
+        email: Joi.string().email().required(),
+        name: Joi.string().required(),
+    }),
+    subject: Joi.string().required(),
+    text: Joi.string().required(),
+    html: Joi.string().required(),
```
Comment (LiyaoJiang1998) [inline backend/services/mailer.js:17]: Just noticed when testing: when `html` is used, the content of `text` seems not used in email.

**Review by LiyaoJiang1998 — CHANGES_REQUESTED** [review]

**Review by LiyaoJiang1998 — APPROVED** [review]

**Comment by LiyaoJiang1998** [comment]
LGTM, just fixed a typo. Tested that sending and validating works.

**Comment by gatilogo** [comment]
> @gatilogo Couldn't add you as one reviewer, please have a look as well.

LGTM

**Review by gza060625 — APPROVED** [review]

**Pull request merged**